### PR TITLE
Atmos sensors

### PR DESCRIPTION
**Pull request by Dunkhan**
Added a set of atmospheric analysis sensors to the integrated electronics module

Consists of 7 new input parts. Pressure, temperature, o2 n2 co2 and phoron levels. The last part is a general atmospheric analyser which has all 6 of those functions in one unit. The multi-purpose part uses triple the complexity and power so it is only worthwhile if you need more than 2 of the output pins.

Please note: There are no changes to the icons, all parts use an existing icon - the one from the advanced medical sensor. I did not feel that this was important to change but someone will probably want to at some point.

### CODE REVIEW

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 1]
```diff
@@ -382,3 +382,236 @@
...
+	complexity = 9
+	inputs = list()
+	outputs = list(
+		"pressure"       = IC_PINTYPE_NUMBER,
+		"temperature" = IC_PINTYPE_NUMBER,
+		"oxygen"         = IC_PINTYPE_NUMBER,
+		"nitrogen"          = IC_PINTYPE_NUMBER,
+		"carbon dioxide"           = IC_PINTYPE_NUMBER,
+		"phoron"           = IC_PINTYPE_NUMBER,
+		"other"           = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Trailing comma, this is akin to:
```dm
outputs = list(
		"pressure"       = IC_PINTYPE_NUMBER,
		"temperature" = IC_PINTYPE_NUMBER,
		"oxygen"         = IC_PINTYPE_NUMBER,
		"nitrogen"          = IC_PINTYPE_NUMBER,
		"carbon dioxide"           = IC_PINTYPE_NUMBER,
		"phoron"           = IC_PINTYPE_NUMBER,
		"other"           = IC_PINTYPE_NUMBER,
		null
	)
```

So just remove the comma from the very last member.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 2]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/pressure_sensor
+	name = "integrated pressure sensor"
+	desc = "A tiny pressure sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"pressure"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma to be removed.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 3]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/temperature_sensor
+	name = "integrated temperature sensor"
+	desc = "A tiny temperature sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"temperature"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 4]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/oxygen_sensor
+	name = "integrated oxygen sensor"
+	desc = "A tiny oxygen sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"oxygen"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 5]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/co2_sensor
+	name = "integrated co2 sensor"
+	desc = "A tiny carbon dioxide sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"co2"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 6]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/nitrogen_sensor
+	name = "integrated nitrogen sensor"
+	desc = "A tiny nitrogen sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"nitrogen"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma here.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 7]
```diff
@@ -382,3 +382,236 @@
...
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/phoron_sensor
+	name = "integrated phoron sensor"
+	desc = "A tiny phoron gas sensor module similar to that found in a PDA atmosphere analyser"
+	icon_state = "medscan_adv"
+	complexity = 3
+	inputs = list()
+	outputs = list(
+		"phoron"       = IC_PINTYPE_NUMBER,
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Comma here.

**Comment by Dunkhan** [comment]
Commas done. Thanks for your help so far

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 8]
```diff
@@ -382,3 +382,236 @@
...
+		set_pin_data(IC_OUTPUT, 4, 0)
+		set_pin_data(IC_OUTPUT, 5, 0)
+		set_pin_data(IC_OUTPUT, 6, 0)
+		set_pin_data(IC_OUTPUT, 7, 0)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/pressure_sensor
+	name = "integrated pressure sensor"
+	desc = "A tiny pressure sensor module similar to that found in a PDA atmosphere analyser"
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Period missing from the end.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 9]
```diff
@@ -382,3 +382,236 @@
...
+	if (total_moles)
+		set_pin_data(IC_OUTPUT, 1, pressure)
+	else
+		set_pin_data(IC_OUTPUT, 1, 0)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/temperature_sensor
+	name = "integrated temperature sensor"
+	desc = "A tiny temperature sensor module similar to that found in a PDA atmosphere analyser"
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Period missing.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 10]
```diff
@@ -382,3 +382,236 @@
...
+	if (total_moles)
+		set_pin_data(IC_OUTPUT, 1, round(environment.temperature-T0C,0.1))
+	else
+		set_pin_data(IC_OUTPUT, 1, -273.15)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/oxygen_sensor
+	name = "integrated oxygen sensor"
+	desc = "A tiny oxygen sensor module similar to that found in a PDA atmosphere analyser"
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Period missing.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 11]
```diff
@@ -382,3 +382,236 @@
...
+	var/total_moles = environment.total_moles
+
+	if (total_moles)
+		set_pin_data(IC_OUTPUT, 1, round(environment.temperature-T0C,0.1))
+	else
+		set_pin_data(IC_OUTPUT, 1, -273.15)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/oxygen_sensor
```
Comment (skull132) [inline code/modules/integrated_electronics/subtypes/input.dm]: Come to think of it. Couldn't you generalize this?

```dm
/obj/item/integrated_circuit/input/gas_sensor
	name = "oxygen gas sensor"
	desc = "A tiny oxygen sensor module similar to that found in a PDA atmosphere analyser."
	icon_state = "medscan_adv"
	complexity = 3
	activators = list("scan" = IC_PINTYPE_PULSE_IN, "on scanned" = IC_PINTYPE_PULSE_OUT)
	spawn_flags = IC_SPAWN_RESEARCH
	origin_tech = list(TECH_ENGINEERING = 3, TECH_DATA = 3)
	power_draw_per_use = 20

	var/gas_name = "oxygen"

/obj/item/integrated_circuit/input/gas_sensor/Initialize()
	. = ..()
	name = "integrated [gas_name] sensor"
	desc = "A tiny [gas_name] sensor module similar to that found in a PDA atmosphere analyser."
	outputs = list(
		gas_name = IC_PINTYPE_NUMBER
	)

/obj/item/integrated_circuit/input/oxygen_sensor/do_work()
	var/turf/T = get_turf(src)
	if(!istype(T)) //Invalid input
		return
	var/datum/gas_mixture/environment = T.return_air()

	var/total_moles = environment.total_moles

	if (total_moles)
		var/gas_level = environment.gas[gas_name]/total_moles
		set_pin_data(IC_OUTPUT, 1, round(gas_level*100,0.1))
	else
		set_pin_data(IC_OUTPUT, 1, 0)
	push_data()
	activate_pin(2)
```

And then just cut others down to like:
```dm
/obj/item/integrated_circuit/input/gas_sensor/co2
	gas_name = "co2"
```

Comment (Dunkhan) [reply code/modules/integrated_electronics/subtypes/input.dm]: hmm I didnt think of that
Will do

**Comment by skull132** [comment]
Apologies for the delay in updating my review, had work.

Looks good, will be merged into dev tomorrrooow.

**Pull request merged**